src/ApplicationInsights.ServiceFabric.Native/Net45/MethodNameProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿namespace Microsoft.ApplicationInsights.ServiceFabric.Remoting.Activities
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+
+    /// <summary>
+    /// Provider for looking up method names by an interface id and a method id. It has methods allowing proxies
+    /// or service objects to be examined, and its interfaces and methods will be added to the map. The ids are
+    /// calculated using the same logic as service fabric runtime, so the method ids used in service remoting can
+    /// be mapped back to method names. 
+    /// </summary>
+    internal class MethodNameProvider : IMethodNameProvider
+    {
+        private Dictionary<int, Dictionary<int, string>> idToMethodNameMap;
+        private HashSet<Type> processedTypes;
+        private object thisLock;
+
+        /// <summary>
+        /// Instantiates the <see cref="MethodNameProvider"/> with the specified remoting factory and retrysettings.
+        /// </summary>
+        public MethodNameProvider()
+        {
+            this.idToMethodNameMap = new Dictionary<int, Dictionary<int, string>>();
+            processedTypes = new HashSet<Type>();
+            thisLock = new object();
+        }
+
+        /// <summary>
+        /// Look up the given interface id and method id and returns the method name.
+        /// This only works for interface types for which it had created proxies.
+        /// </summary>
+        /// <param name="interfaceId"></param>
+        /// <param name="methodId"></param>
+        /// <returns></returns>
+        public string GetMethodName(int interfaceId, int methodId)
+        {
+            lock (thisLock)
+            {
+                if (this.idToMethodNameMap.TryGetValue(interfaceId, out Dictionary<int, string> methodMap))
+                {
+                    if (methodMap.TryGetValue(methodId, out string methodName))
+                    {
+                        return methodName;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Adds methods for the given interfaces which derives from the given base interface type.
+        /// </summary>
+        /// <param name="interfaces">The interfaces to be added.</param>
+        /// <param name="baseInterfaceType">The base interface type as a filter where only interfaces that derives for this base interface type are added.</param>
+        public void AddMethodsForProxyOrService(IEnumerable<Type> interfaces, Type baseInterfaceType)
```

Comment:
i might be wrong, but it seems AddMethodsForProxyOrService is only called during client/handler setup, while the GetMethodName is only called when the RPC call is in progress.
Am I wrong and there is a concurrency? Ie. is there a way to make GetMethodName lock (and concurrent-dictionary) free?

Reply:
You are correct about when they are getting used, however, on the client side, those can happen simultaneously. Since the MethodNameProvider is owned by the proxy factory object, the app can use the same factory to create multiple proxies, and these proxies can send requests at different times, which creates concurrencies..

On the handler side, since the handler owns the MethodNameProvider completely and it adds to the map only at construction time, what you said is correct. I will see if I can optimize it.